File: backend/src/motion/agents/soap_agents/message_types.py
"""
Message type definitions for structured frontend communication.
"""

from enum import Enum
from typing import List, Dict, Any, Optional
from datetime import datetime
import json


class MessageType(str, Enum):
    """Enumeration of message types for frontend rendering."""
    CHAT_MESSAGE = "chat_message"
    SOAP_DRAFT = "soap_draft"
    EXERCISE_SELECTION = "exercise_selection"
    FINAL_REPORT = "final_report"
    CLARIFICATION = "clarification_needed"
    ERROR = "error"


class StructuredMessage:
    """Base class for structured messages to frontend."""
    
    def __init__(self, message_type: MessageType, **kwargs):
        self.type = message_type
        self.timestamp = datetime.now().isoformat()
        self.data = kwargs
    
    def to_json(self) -> str:
        """Convert message to JSON string for transmission."""
        return json.dumps({
            "type": self.type,
            "timestamp": self.timestamp,
            **self.data
        }, indent=2)


class ChatMessage(StructuredMessage):
    """Message containing general chat response."""
    
    def __init__(self, content: str):
        super().__init__(
            MessageType.CHAT_MESSAGE,
            content=content
        )


class SoapDraftMessage(StructuredMessage):
    """Message containing initial SOAP report draft in structured format."""
    
    def __init__(self, soap_report: Dict[str, Any]):
        super().__init__(
            MessageType.SOAP_DRAFT,
            soap_report=soap_report
        )


class ExerciseSelectionMessage(StructuredMessage):
    """Message prompting user to select exercise images for multiple exercises."""
    
    def __init__(self, exercises: List[Dict[str, Any]]):
        super().__init__(
            MessageType.EXERCISE_SELECTION,
            exercises=exercises,
            requires_selection=True
        )


class FinalReportMessage(StructuredMessage):
    """Message containing final SOAP report with selected images."""
    
    def __init__(self, soap_report: Dict[str, Any], selected_images: List[str]):
        super().__init__(
            MessageType.FINAL_REPORT,
            soap_report=soap_report,
            selected_images=selected_images,
            ready_for_pdf=True
        )


class ClarificationMessage(StructuredMessage):
    """Message requesting clarification from user."""
    
    def __init__(self, questions: List[str], original_content: str):
        super().__init__(
            MessageType.CLARIFICATION,
            questions=questions,
            original_content=original_content
        )


class ErrorMessage(StructuredMessage):
    """Message indicating an error occurred."""
    
    def __init__(self, error: str, details: Optional[str] = None):
        super().__init__(
            MessageType.ERROR,
            error=error,
            details=details
        )


def create_chat_message(content: str) -> str:
    """Helper function to create chat message JSON."""
    return ChatMessage(content).to_json()


def create_soap_draft_message(soap_report: Dict[str, Any]) -> str:
    """Helper function to create SOAP draft message JSON."""
    return SoapDraftMessage(soap_report).to_json()


def create_exercise_selection_message(exercises: List[Dict[str, Any]]) -> str:
    """Helper function to create exercise selection message JSON."""
    return ExerciseSelectionMessage(exercises).to_json()


def create_final_report_message(soap_report: Dict[str, Any], selected_images: List[str]) -> str:
    """Helper function to create final report message JSON."""
    return FinalReportMessage(soap_report, selected_images).to_json()


def create_clarification_message(questions: List[str], original_content: str) -> str:
    """Helper function to create clarification message JSON."""
    return ClarificationMessage(questions, original_content).to_json()


def create_error_message(error: str, details: Optional[str] = None) -> str:
    """Helper function to create error message JSON."""
    return ErrorMessage(error, details).to_json()


# Helper functions for creating structured SOAP data

def create_soap_item(item_type: str, content: str, emphasis: Optional[str] = None, sub_items: Optional[List[Dict[str, Any]]] = None) -> Dict[str, Any]:
    """Create a SOAP item with proper structure."""
    return {
        "type": item_type,
        "content": content,
        "emphasis": emphasis,
        "sub_items": sub_items
    }

def create_soap_section(title: str, items: List[Dict[str, Any]]) -> Dict[str, Any]:
    """Create a SOAP section with proper structure."""
    return {
        "title": title,
        "items": items
    }

def create_patient_info(name: Optional[str] = None, age: Optional[str] = None, 
                       condition: Optional[str] = None, session_date: Optional[str] = None) -> Dict[str, Any]:
    """Create patient info structure."""
    return {
        "name": name,
        "age": age,
        "condition": condition,
        "session_date": session_date
    }

def create_soap_report(patient_info: Optional[Dict[str, Any]], subjective: Dict[str, Any], 
                      objective: Dict[str, Any], assessment: Dict[str, Any], 
                      plan: Dict[str, Any], timestamp: str) -> Dict[str, Any]:
    """Create complete SOAP report structure."""
    return {
        "patient_info": patient_info,
        "subjective": subjective,
        "objective": objective,
        "assessment": assessment,
        "plan": plan,
        "timestamp": timestamp
    }

def create_exercise_with_images(exercise_id: str, name: str, description: str, 
                               search_results: List[Dict[str, Any]]) -> Dict[str, Any]:
    """Create exercise structure with processed images."""
    processed_images = []
    for i, img in enumerate(search_results):
        processed_images.append({
            "id": f"img_{exercise_id}_{i}",
            "url": img.get("url", ""),
            "name": img.get("name", f"{name} illustration {i+1}"),
            "selected": False
        })
    
    return {
        "id": exercise_id,
        "name": name,
        "description": description,
        "images": processed_images
    }

File: ios/MotionByAiselu/MotionByAiselu/Models/MessageTypes.swift
import Foundation

// MARK: - Message Types
enum MessageType: String, Codable {
    case chatMessage = "chat_message"
    case soapDraft = "soap_draft"
    case exerciseSelection = "exercise_selection"
    case finalReport = "final_report"
    case clarificationNeeded = "clarification_needed"
    case error = "error"
}

// MARK: - Main Message Structure
struct StructuredMessage: Codable, Identifiable {
    let id = UUID()
    let type: MessageType
    let timestamp: String
    
    // Common properties (optional based on message type)
    let content: String?
    let format: String?
    
    // Exercise selection specific
    let exerciseName: String?
    let exerciseDescription: String?
    let images: [ExerciseImage]?
    let exercises: [Exercise]? // New: multiple exercises support
    let requiresSelection: Bool?
    
    // Final report specific
    let selectedImages: [String]?
    let readyForPdf: Bool?
    
    // Structured SOAP report
    let soapReport: SOAPReport?
    
    // Clarification specific
    let questions: [String]?
    let originalContent: String?
    
    // Error specific
    let error: String?
    let details: String?
    
    private enum CodingKeys: String, CodingKey {
        case type, timestamp, content, format
        case exerciseName = "exercise_name"
        case exerciseDescription = "exercise_description"
        case images
        case exercises
        case requiresSelection = "requires_selection"
        case selectedImages = "selected_images"
        case readyForPdf = "ready_for_pdf"
        case soapReport = "soap_report"
        case questions
        case originalContent = "original_content"
        case error, details
    }
}

// MARK: - Exercise
struct Exercise: Codable, Identifiable {
    let id: String
    let name: String
    let description: String
    let images: [ExerciseImage]
    
    init(id: String, name: String, description: String, images: [ExerciseImage]) {
        self.id = id
        self.name = name
        self.description = description
        self.images = images
    }
}

// MARK: - Exercise Image
struct ExerciseImage: Codable, Identifiable {
    let id: String
    let url: String
    let name: String
    var selected: Bool
    
    init(id: String, url: String, name: String, selected: Bool = false) {
        self.id = id
        self.url = url
        self.name = name
        self.selected = selected
    }
}

// MARK: - API Request Types
struct AgentRunRequest: Codable {
    let appName: String
    let userId: String
    let sessionId: String
    let newMessage: MessageContent
    let streaming: Bool
    
    private enum CodingKeys: String, CodingKey {
        case appName = "app_name"
        case userId = "user_id"
        case sessionId = "session_id"
        case newMessage = "new_message"
        case streaming
    }
}

struct MessageContent: Codable {
    let role: String
    let parts: [MessagePart]
    
    init(text: String) {
        self.role = "user"
        self.parts = [MessagePart(text: text)]
    }
}

struct MessagePart: Codable {
    let text: String
}

// MARK: - API Response Types
struct APIResponse: Codable {
    let events: [AgentEvent]?
    let error: String?
}

struct AgentEvent: Codable {
    let id: String?
    let author: String?
    let content: EventContent?
    let timestamp: Double?
    let invocationId: String?
    
    private enum CodingKeys: String, CodingKey {
        case id, author, content, timestamp
        case invocationId = "invocationId"
    }
}

struct EventContent: Codable {
    let parts: [MessagePart]?
    let role: String?
}

// MARK: - Helper for dynamic JSON
struct AnyCodable: Codable {
    let value: Any
    
    init(_ value: Any) {
        self.value = value
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        
        if let value = try? container.decode(String.self) {
            self.value = value
        } else if let value = try? container.decode(Int.self) {
            self.value = value
        } else if let value = try? container.decode(Double.self) {
            self.value = value
        } else if let value = try? container.decode(Bool.self) {
            self.value = value
        } else if let value = try? container.decode([String: AnyCodable].self) {
            self.value = value
        } else if let value = try? container.decode([AnyCodable].self) {
            self.value = value
        } else {
            throw DecodingError.typeMismatch(AnyCodable.self, DecodingError.Context(codingPath: decoder.codingPath, debugDescription: "Unsupported type"))
        }
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        
        switch value {
        case let string as String:
            try container.encode(string)
        case let int as Int:
            try container.encode(int)
        case let double as Double:
            try container.encode(double)
        case let bool as Bool:
            try container.encode(bool)
        case let dict as [String: AnyCodable]:
            try container.encode(dict)
        case let array as [AnyCodable]:
            try container.encode(array)
        default:
            throw EncodingError.invalidValue(value, EncodingError.Context(codingPath: encoder.codingPath, debugDescription: "Unsupported type"))
        }
    }
}

// MARK: - Image Selection Request
struct ImageSelectionRequest: Codable {
    let selectedImageIds: [String]
    let messageType: String
    
    private enum CodingKeys: String, CodingKey {
        case selectedImageIds = "selected_image_ids"
        case messageType = "message_type"
    }
    
    init(selectedImageIds: [String]) {
        self.selectedImageIds = selectedImageIds
        self.messageType = "image_selection"
    }
}


// MARK: - Simplified SOAP Report
struct SOAPReport: Codable, Identifiable {
    let id = UUID()
    let patientName: String?
    let patientAge: String?
    let condition: String?
    let sessionDate: String?
    let subjective: String
    let objective: String
    let assessment: String
    let plan: String
    let exercises: [SimpleExercise]
    let timestamp: String?
    
    private enum CodingKeys: String, CodingKey {
        case patientName = "patient_name"
        case patientAge = "patient_age"
        case condition
        case sessionDate = "session_date"
        case subjective, objective, assessment, plan, exercises, timestamp
    }
}

struct SimpleExercise: Codable, Identifiable {
    let id = UUID()
    let name: String
    let description: String
    let selectedImage: String?
    
    private enum CodingKeys: String, CodingKey {
        case name, description
        case selectedImage = "selected_image"
    }
}